Editorial review: Document FileSystemObserver (#38270)
mdn/content: files/en-us/web/api/filesystemobserver/observe/index.md

---
title: "FileSystemObserver: observe() method"
short-title: observe()
slug: Web/API/FileSystemObserver/observe
page-type: web-api-instance-method
browser-compat: api.FileSystemObserver.observe
---

{{securecontext_header}}{{APIRef("File System API")}}

The **`observe()`** method of the

> {{domxref("FileSystemObserver")}} interface starts the observer observing changes to a given file or directory.

Reviewer: ```suggestion
{{domxref("FileSystemObserver")}} interface asks the observer to start observing changes to a given file or directory.
```

I dunno, "starts the observer observing" seemed a bit hard to read.

## Syntax

```js-nolint
observe(handle)
observe(handle, options)
```

### Parameters

- `handle`

>   - : The handle of the file system entry representing the file or directory to observe. This can be a {{domxref("FileSystemFileHandle")}} or {{domxref("FileSystemDirectoryHandle")}} for the user-observable file system, or a {{domxref("FileSystemSyncAccessHandle")}} for the [Origin Private File System](/en-US/docs/Web/API/File_System_API/Origin_private_file_system) (OPFS).

Reviewer 2: ```suggestion
  - : The handle of the file system entry representing the file or directory to observe. This can be a {{domxref("FileSystemFileHandle")}} or {{domxref("FileSystemDirectoryHandle")}} for the user-observable file system and the [Origin Private File System](/en-US/docs/Web/API/File_System_API/Origin_private_file_system) (OPFS), or a {{domxref("FileSystemSyncAccessHandle")}} for the OPFS.
```

Author: fixed

>   - : The handle of the file system entry representing the file or directory to observe. This can be a {{domxref("FileSystemFileHandle")}} or {{domxref("FileSystemDirectoryHandle")}} for the user-observable file system and the [Origin Private File System](/en-US/docs/Web/API/File_System_API/Origin_private_file_system) (OPFS), or a {{domxref("FileSystemSyncAccessHandle")}} for the OPFS.

Reviewer: Same comment as before:

```suggestion
  - : The handle of the file system entry representing the file or directory to observe.
      - For the user-observable file system, this can be a {{domxref("FileSystemFileHandle")}} or a {{domxref("FileSystemDirectoryHandle")}}.
      - For the [Origin Private File System](/en-US/docs/Web/API/File_System_API/Origin_private_file_system) (OPFS), it can be or a a {{domxref("FileSystemFileHandle")}}, a {{domxref("FileSystemDirectoryHandle")}}, or a {{domxref("FileSystemSyncAccessHandle")}}.
```

Author: updated

- `options` {{optional_inline}}

  - : An object specifying options for the `observe()` call. This can contain the following properties:

    - `recursive`

>       - : A boolean specifying whether you want to observe changes to a directory recursively. If set to `true`, changes are observed in the directory itself and all contained subdirectories and files. If set to `false`, changes are only observed in the directory itself and directly contained files (that is, not files in subdirectories). Defaults to `false`.

Reviewer: ```suggestion
      - : A boolean specifying whether you want to observe changes to a directory recursively. If set to `true`, changes are observed in the directory itself and all contained subdirectories and files. If set to `false`, changes are only observed in the directory itself and directly contained files (that is, files in subdirectories are excluded). Defaults to `false`.
```

        This property has no effect if `handle` represents a file.

### Return value

A {{jsxref("Promise")}} that resolves to {{jsxref('undefined')}}.

### Exceptions

- `NotFoundError` {{domxref("DOMException")}}
  - : Thrown if the file or directory represented by `handle` could not be found.

## Examples

### Observe a file or directory

> Assuming a `FileSystemObserver` instance is available, you can start observing changes to a file system entry by calling the `observe()` method on it, passing it the handle of the entry to observe. This can be a {{domxref("FileSystemFileHandle")}} or {{domxref("FileSystemDirectoryHandle")}} for the user-observable file system:

Reviewer 2: ```suggestion
Assuming a `FileSystemObserver` instance is available, you can start observing changes to a file system entry by calling the `observe()` method on it, passing it the handle of the entry to observe. This can be a {{domxref("FileSystemFileHandle")}} or {{domxref("FileSystemDirectoryHandle")}} for the user-observable file system and the the [Origin Private File System](/en-US/docs/Web/API/File_System_API/Origin_private_file_system) (OPFS):
```

Author: fixed

> Assuming a `FileSystemObserver` instance is available, you can start observing changes to a file system entry by calling the `observe()` method on it, passing it the handle of the entry to observe. This can be a {{domxref("FileSystemFileHandle")}} or {{domxref("FileSystemDirectoryHandle")}} for the user-observable file system and the [Origin Private File System](/en-US/docs/Web/API/File_System_API/Origin_private_file_system) (OPFS):

Reviewer: As before, I think explaining the different use cases is more helpful in an example than duplicating the types you can pass.

Author: updated, same as the other page.

```js
// Observe a file
async function observeFile() {
  const fileHandle = await window.showSaveFilePicker();

  await observer.observe(fileHandle);
}

// Observe a directory
async function observeDirectory() {
  const directoryHandle = await window.showDirectoryPicker();

  await observer.observe(directoryHandle);
}
```

or a {{domxref("FileSystemSyncAccessHandle")}} for the [Origin Private File System](/en-US/docs/Web/API/File_System_API/Origin_private_file_system) (OPFS):

```js
// Observe an OPFS file system entry
async function observeOPFSFile() {
  const root = await navigator.storage.getDirectory();
  const draftHandle = await root.getFileHandle("draft.txt", { create: true });

>   const opfsHandle = await draftHandle.createSyncAccessHandle();
> 
>   await observer.observe(opfsHandle);

Reviewer 2: ```suggestion
  const syncHandle = await draftHandle.createSyncAccessHandle();

  await observer.observe(syncHandle);
```

Author: fixed
